File: src/auth/store.ts
/**
 * Credential persistence: `credentials.json` in `dataDir` (mode 0600).
 * One `Credential[]` array per provider id.
 *
 * Port of `auth/store.rs`. Merge dedup semantics are EXACTLY the Rust
 * `merge()`: a credential with an identity replaces any existing one with the
 * same identity; a credential without an identity replaces the lone
 * identity-less credential; distinct identities accumulate (multi-account).
 */
import { chmodSync, existsSync, readFileSync, statSync, writeFileSync } from "node:fs";
import { dirname } from "node:path";
import { mkdirSync } from "node:fs";
import { credentialsFile } from "../config/paths.ts";
import type { Credential } from "./types.ts";
import { identity } from "./types.ts";
import { SUPPORTED_PROVIDERS } from "../usage/types.ts";

export function load(dataDir: string, provider: string): Credential[] {
  const file = credentialsFile(dataDir);
  if (!existsSync(file)) return [];
  const all = readAll(file);
  if (!all) return [];
  warnIfInsecurePermissions(file);
  const entry = all[provider];
  return entry ? entry.slice() : [];
}

export function save(dataDir: string, provider: string, creds: Credential[]): void {
  const file = credentialsFile(dataDir);
  const all = existsSync(file) ? readAll(file) ?? {} : {};
  if (creds.length === 0) {
    delete all[provider];
  } else {
    all[provider] = creds;
  }
  writeAtomic(file, all);
  chmodSync(file, 0o600);
}

export function upsert(dataDir: string, provider: string, cred: Credential): void {
  const creds = load(dataDir, provider);
  merge(creds, cred);
  save(dataDir, provider, creds);
}

export function removeAccount(dataDir: string, provider: string, accountIdentity: string): boolean {
  const creds = load(dataDir, provider);
  const idx = creds.findIndex((c) => identity(c) === accountIdentity);
  if (idx < 0) return false;
  creds.splice(idx, 1);
  save(dataDir, provider, creds);
  return true;
}

export function allProviders(dataDir: string): string[] {
  const file = credentialsFile(dataDir);
  if (!existsSync(file)) return [];
  const all = readAll(file);
  if (!all) return [];
  const supported = new Set(SUPPORTED_PROVIDERS);
  return Object.keys(all).filter((k) => supported.has(k) && Array.isArray(all[k]) && all[k].length > 0);
}

/**
 * Pure dedup (exported for tests). Mutates `creds` in place.
 * Mirrors Rust `merge()`: same identity replaces; identity-less replaces the
 * lone identity-less credential; distinct identities accumulate.
 */
export function merge(creds: Credential[], cred: Credential): void {
  const id = identity(cred);
  if (id != null) {
    for (let i = 0; i < creds.length; i++) {
      if (identity(creds[i]) === id) {
        creds[i] = cred;
        return;
      }
    }
    creds.push(cred);
    return;
  }
  for (let i = 0; i < creds.length; i++) {
    if (identity(creds[i]) == null) {
      creds[i] = cred;
      return;
    }
  }
  creds.push(cred);
}

function readAll(file: string): Record<string, Credential[]> | null {
  try {
    const txt = readFileSync(file, "utf8");
    if (!txt.trim()) return {};
    const parsed = JSON.parse(txt);
    if (!parsed || typeof parsed !== "object" || Array.isArray(parsed)) return {};
    return parsed as Record<string, Credential[]>;
  } catch {
    return null;
  }
}

function writeAtomic(file: string, all: Record<string, Credential[]>): void {
  mkdirSync(dirname(file), { recursive: true });
  writeFileSync(file, JSON.stringify(all, null, 2));
}

function warnIfInsecurePermissions(file: string): void {
  try {
    const mode = statSync(file).mode & 0o777;
    if (mode & 0o077) {
      process.stderr.write(
        `warning: ${file} is group/world readable (mode 0o${mode.toString(8)}); chmod 600 recommended\n`,
      );
    }
  } catch {
    // ignore
  }
}

File: src/auth/store.rs
//! Persistence for atop credentials: one JSON array of {@link Credential} per
//! provider id, kept in the `secrets` table of `atop.db` (AES-256-GCM).
use anyhow::Result;

use super::Credential;
use crate::db::Db;

pub fn load(db: &Db, provider: &str) -> Result<Vec<Credential>> {
    match db.get_secret(provider)? {
        Some(s) if !s.trim().is_empty() => Ok(serde_json::from_str(&s).unwrap_or_default()),
        _ => Ok(Vec::new()),
    }
}

pub fn save(db: &Db, provider: &str, creds: &[Credential]) -> Result<()> {
    if creds.is_empty() {
        return db.delete_secret(provider);
    }
    db.set_secret(provider, &serde_json::to_string(creds)?)
}

/// Pure dedup: insert `cred`, replacing any existing credential with the same
/// identity (or the lone identity-less credential for single-account providers).
pub(crate) fn merge(creds: &mut Vec<Credential>, cred: Credential) {
    match cred.identity() {
        Some(id) => {
            if let Some(slot) = creds.iter_mut().find(|c| c.identity().as_deref() == Some(id.as_str())) {
                *slot = cred;
            } else {
                creds.push(cred);
            }
        }
        None => {
            if let Some(slot) = creds.iter_mut().find(|c| c.identity().is_none()) {
                *slot = cred;
            } else {
                creds.push(cred);
            }
        }
    }
}

/// Insert `cred` into the provider's stored credentials, then persist.
pub fn upsert(db: &Db, provider: &str, cred: Credential) -> Result<()> {
    let mut creds = load(db, provider)?;
    merge(&mut creds, cred);
    save(db, provider, &creds)
}

/// Providers that currently have stored credentials in atop's secret store.
pub fn all_providers(db: &Db) -> anyhow::Result<Vec<&'static str>> {
    let mut out = Vec::new();
    for p in crate::usage::supported() {
        if !load(db, p)?.is_empty() {
            out.push(*p);
        }
    }
    Ok(out)
}

pub fn remove(db: &Db, provider: &str) -> Result<()> {
    db.delete_secret(provider)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::auth::{ApiKeyCred, OAuthCred};

    fn api(key: &str) -> Credential {
        Credential::ApiKey(ApiKeyCred { key: key.into(), account: None, enterprise_url: None })
    }
    fn oauth(email: &str) -> Credential {
        Credential::Oauth(OAuthCred {
            access: "a".into(),
            refresh: Some("r".into()),
            expires: Some(0),
            account_id: None,
            email: Some(email.into()),
            project_id: None,
            enterprise_url: None,
        })
    }

    #[test]
    fn merge_dedupes_identityless_and_keeps_distinct_identities() {
        let mut creds: Vec<Credential> = Vec::new();
        merge(&mut creds, api("k1"));
        merge(&mut creds, api("k2"));
        assert_eq!(creds.len(), 1, "identity-less api keys collapse to one");
        match &creds[0] {
            Credential::ApiKey(c) => assert_eq!(c.key, "k2"),
            _ => panic!("expected api key"),
        }

        let mut accts: Vec<Credential> = Vec::new();
        merge(&mut accts, oauth("a@x.com"));
        merge(&mut accts, oauth("b@x.com"));
        merge(&mut accts, oauth("a@x.com"));
        assert_eq!(accts.len(), 2, "two distinct emails kept; duplicate replaced");
    }

    #[test]
    fn json_roundtrip_preserves_kind() {
        let creds = vec![api("k1"), oauth("a@x.com")];
        let s = serde_json::to_string(&creds).unwrap();
        let back: Vec<Credential> = serde_json::from_str(&s).unwrap();
        assert!(matches!(back[0], Credential::ApiKey(_)));
        assert!(matches!(back[1], Credential::Oauth(_)));
    }
}
